Automate plotting

mcwitt opened the pull request:
Closes #33, #48 

### Run-level plots

#### Work distributions for solvent and complex (`RUN0`)
![RUN0](https://user-images.githubusercontent.com/319411/91625862-86be3a00-e95f-11ea-8d63-fba01070f7e0.png)

#### Convergence of relative free energies by GEN (`RUN0-convergence`)
![RUN0-convergence](https://user-images.githubusercontent.com/319411/91741886-349d3480-eb6a-11ea-88f6-666553fe5e98.png)

### Summary plots

#### Distribution of relative binding free energies (`relative_fe_dist`)
![relative_fe_dist](https://user-images.githubusercontent.com/319411/91737471-fb61c600-eb63-11ea-8de3-25175a9c2e94.png)

#### Cumulative distribution of relative free binding energies (`cumulative_fe_dist`)
![cumulative_fe_hist](https://user-images.githubusercontent.com/319411/91624754-dd734600-e956-11ea-9bf0-ed739d67aaed.png)

#### TODO
- [x] Add convergence plot (requires #36)
- [x] Add missing docstrings

mcwitt commented on covid_moonshot/analysis/fah_plotting.py:
```diff
@@ -76,49 +60,73 @@ def plot_relative_distribution(relative_fes, bins=100, title='Relative affinity'
-    _produce_plot(filename)
 
 
 def plot_convergence(results, n_gens=3, title=None):
```
@hannahbrucemacdonald @jchodera it looks like this plot relies on results at the GEN level, which we are not yet computing in the analysis pipeline. Is it worth adding this analysis, or better to skip this plot for now?

jchodera replied:
Let's break out the GEN-level info into a separate issue to first discuss how to represent work values can be organized by GEN. That's something that will be important as a sanity check.

jchodera commented:
@mcwitt Can you summarize what gets plotted (with examples) and where it ends up?

jchodera commented on covid_moonshot/analysis/fah_plotting.py:
```diff
@@ -1,154 +1,191 @@
+import matplotlib.figure
```
Not necessary to do this now, but in general, we recommend lazy imports inside methods that use external packages so that if we reuse components of this in other tools, (1) it doesn't take forever to import, and (2) we can have optional dependencies for certain functionality. This also helps enormously when using readthedocs, since we don't need to have all dependencies installed just for sphinx to inspect the docstrings and generate API docs.

mcwitt replied:
Makes sense. I'd been doing this mostly with modules that aren't in PyPI / in `install_requires`, but I can be more aggressive about this.

jchodera commented on covid_moonshot/analysis/fah_plotting.py:
```diff
@@ -1,154 +1,191 @@
-        logging.info(f"Writing {fname}")
-        plt.savefig(fname)
-
+def plot_single_work_distribution(
```
Missing docstring?

mcwitt replied:
Added now (mostly 74c4814)

jchodera commented on covid_moonshot/analysis/fah_plotting.py:
```diff
@@ -1,154 +1,191 @@
-    fig, (ax1, ax2) = plt.subplots(ncols=2, nrows=1, figsize=(7.5, 3.25))
-    plot_single_work_distribution(f1, r1, ax1, phases[0])
-    plot_single_work_distribution(f2, r2, ax2, phases[1])
+def plot_two_work_distribution(
```
Missing docstring?

mcwitt replied:
74c4814

jchodera commented on covid_moonshot/analysis/fah_plotting.py:
```diff
@@ -1,154 +1,191 @@
-                                  filename='Cumulative distribution'):
+
+
+def plot_convergence(
```
Missing docstring?

mcwitt replied:
74c4814

jchodera commented on covid_moonshot/analysis/fah_plotting.py:
```diff
@@ -1,154 +1,191 @@
+def plot_convergence(
+    gens: List[int],
+    complex_delta_fs: List[float],
+    complex_delta_f_errs: List[float],
```
Should we move toward a data structure that encapsulates the phase (complex, solvent) and overall (binding) free energies and uncertainties in a single object?

mcwitt replied:
Yes, this already exists as a matter of fact (the `RunAnalysis` object). I intentionally kept the interface to the plotting functions low-level (i.e. not using the types in `covid_moonshot.core`) in case they would be useful to call from other code where we don't want the full baggage of the object hierarchy. I could go either way though, honestly.

jchodera commented on covid_moonshot/analysis/fah_plotting.py:
```diff
@@ -179,19 +216,83 @@ def plot_cumulative_distributions(results, minimum=None, maximum=5, cmap='PiYG',
+    plt.ylabel("Cumulative $N$ ligands")
+
+
+def get_plot_filename(path: str, name: str, file_format: str) -> str:
```
Missing docstring?

mcwitt replied:
Addressed by 41e9e6a, 6717294

jchodera commented on covid_moonshot/analysis/fah_plotting.py:
```diff
@@ -179,19 +216,83 @@ def plot_cumulative_distributions(results, minimum=None, maximum=5, cmap='PiYG',
+    return os.path.join(path, os.extsep.join([name, file_format]))
+
+
+def save_run_level_plots(
```
Missing docstring?

mcwitt replied:
Added in 74c4814

jchodera commented on covid_moonshot/analysis/fah_plotting.py:
```diff
@@ -179,19 +216,83 @@ def plot_cumulative_distributions(results, minimum=None, maximum=5, cmap='PiYG',
+    plt.savefig(get_plot_filename(path, f"run{run}-convergence", file_format))
+
+
+def save_summary_plots(
```
Missing docstring?

mcwitt replied:
Added in 41e9e6a

jchodera commented on covid_moonshot/lib.py:
```diff
@@ -123,6 +125,14 @@ def analyze_run(
         except ValueError as e:
             raise ValueError(f"Failed to save structures for complex: {e}")
 
+    save_run_level_plots(
```
Missing docstring?

mcwitt replied:
Addressed by 41e9e6a, 6717294

mcwitt replied:
Added in 74c4814

jchodera commented on setup.py:
```diff
@@ -12,7 +12,9 @@
         "joblib>=0.16",
```
Should we also have an `environment.yml` to make it easier to configure the run environment? The new MolSSI Cookiecutter requires this, so we can wait until we migrate.

mcwitt replied:
Definitely on board with this! I'm having an issue with conda currently (my version doesn't have the `--from-history` option for `conda env export`, and the newer versions seem broken for my OS). I'll plan to add it to a later PR, or when we do the migration.

mcwitt commented:
@jchodera I added a brief summary of the run-level and summary plots to the docstrings of the relevant functions in 68c787b. Also added example images to the PR description.

jchodera commented:
This is amazing!

For the work distributions:
* Can we use "$k_B T$ instead of having "300 K" in there?
* Let's add a vertical black solid or dashed line for the free energy estimate from all runs, if it's easy to do that. It should appear near the intersection of the normalized work distributions, but is a useful visual check/guide.

Otherwise, this looks _fantastic_!

jchodera commented on covid_moonshot/analysis/plots.py:
```diff
@@ -0,0 +1,436 @@
+    ----------
+    ax : AxesSubplot
+       Axes on which to draw the plot
+    forward_works : list of float
```
Apologies for just realizing this: Would be good to document units for work or energy values if they are floats! You can just say "(in kT)" for all these.

mcwitt replied:
Cool, I can add this. My convention so far is to assume kT and specify only when we differ from that, but explicit is better than implicit!

mcwitt replied:
Will be cool to transition to `pint` or something similar!

mcwitt replied:
Annotated docstrings with `(in kT)` for now: 7b59b1e

mcwitt commented:
@jchodera 
> Let's add a vertical black solid or dashed line for the free energy estimate from all runs
I'm a bit confused here. Do you mean the binding free energy estimate for the current run? (We're not calculating any aggregates over all runs in the pipeline)

jchodera commented:
Whoops! Each phase (complex, solvent) has a BAR free energy estimate, and when plotted as a vertical line at the free energy estimate on those work distribution plots, should appear near the intersection of distributions. I can draw a sketch if that would be helpful!

mcwitt commented:
@jchodera makes sense! Added in 8493d31 and updated the example image.

jchodera commented:
💯

jchodera commented on covid_moonshot/analysis/plots.py:
```diff
@@ -0,0 +1,448 @@
+
+    Parameters
+    ----------
+    affinities : list of float
```
What units are these supposed to be in?
Let's also be a bit careful to call then "delta_f" or "free_energies" since the relationship between the free energy and affinity is
```
affinity = exp(-free_energy / kT) * (1 Molar)
```
We generally don't want to use affinities in our reports for internal use.

mcwitt replied:
Ah! I wasn't aware of this convention. Makes sense. I replaced all occurrences of "affinity" with "delta_f" or "free energy" in 9eddd3d, 56cfc0b

jchodera commented:
I've activated this branch on `aws3` to test the plotting pipeline, and there appears to be a discrepancy between the PDF/CDF plots and the summarized `ligands.pdf`. The CDF shows the best ligands are < -2 kcal/mol:
![image](https://user-images.githubusercontent.com/3656088/91651734-1ab60180-ea45-11ea-8918-ef1bb4109314.png)
But the `ligands.pdf` file the `postprocess.py` script (which assumes the JSON outputs binding free energies in kT and converts to kcal/mol) generates shows no ligands < -2 kcal/mol:
![image](https://user-images.githubusercontent.com/3656088/91651747-2ef9fe80-ea45-11ea-8891-2737a4030454.png)
We should be a bit careful with units to use kT for all serialization but we can plot with kcal/mol.

mcwitt commented:
> I've activated this branch on aws3 to test the plotting pipeline, and there appears to be a discrepancy between the PDF/CDF plots 

@jchodera I think this is due to the comparison of values in kT and kcal/mol, fixed in 54a517d. I'll try to reproduce your example to make sure.

mcwitt commented:
> We should be a bit careful with units to use kT for all serialization but we can plot with kcal/mol.

The convention I'm using now is:
- All free energies are in `kT`, unless the name is explicitly suffixed with `_kcal`
- Values in kcal/mol only exist in the plotting code `analysis.plots` (everything we serialize is kT)

I'll look at adding `pint` or something similar soon. It seems really useful to be able to catch errors like the one fixed in 
54a517d at runtime (or even better, with a red underline in the editor, if there is mypy support)

jchodera commented:
Whoops, not sure why these commits showed up as being from @hannahbrucemacdonald, but 
17afc59 should fix the off-by-one problem (#46)!

hannahbrucemacdonald commented:
Looks good to me! Any reason that the convergence plots are deltaF rather than deltaG?

mcwitt commented:
> Looks good to me! Any reason that the convergence plots are deltaF rather than deltaG?

No good reason! Delta G is more appropriate, correct? I can change it.
